Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm

```diff
@@ -351,20 +351,26 @@
 {
     auto *ruleIDsToDelete = @[ ];
     if (ruleIDsToDeleteVector) {
-        ruleIDsToDelete = createNSArray(ruleIDsToDeleteVector.value(), [](double ruleID) {
+        ruleIDsToDelete = createNSArray(ruleIDsToDeleteVector.value(), [this, protectedThis = Ref { *this }](double ruleID) {
+            if (!m_dynamicRulesIDs.contains(ruleID))
+                return (NSNumber *)nil;
             return @(ruleID);
         }).get();
     }
 
     NSArray *rulesToAdd = rulesToAddJSON ? parseJSON(rulesToAddJSON.value(), JSONOptions::FragmentsAllowed) : @[ ];
 
-    // FIXME: Make sure that adding these rules won't get us over the maximum number of dynamic + session rules.
-
     if (!ruleIDsToDelete.count && !rulesToAdd.count) {
         completionHandler(std::nullopt);
         return;
     }
 
+    double updatedDynamicRulesCount = m_dynamicRulesIDs.size() + rulesToAdd.count - ruleIDsToDelete.count;
+    if (updatedDynamicRulesCount + m_sessionRulesIDs.size() > webExtensionDeclarativeNetRequestMaximumNumberOfDynamicAndSessionRules) {
+        completionHandler(toErrorString(@"declarativeNetRequest.updateDynamicRules()", nil, @"Failed to add dynamic rules. Maximum number of dynamic and session rules exceeded."));
+        return;
+    }
+
     updateDeclarativeNetRequestRulesInStorage(declarativeNetRequestDynamicRulesStore(), @"dynamic", rulesToAdd, ruleIDsToDelete, WTFMove(completionHandler));
```

Comment:
You need code that removes the deleted IDs from `m_dynamicRulesIDs`.

Reply:
This is taken care of by the fact that we are replacing the whole vector.